src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,155 @@
+        public int Dimensions => Point ? Coordinates.Count() : Coordinates.Count() / 2;
+
+        /// <summary>
+        /// Is the cube a point.
+        /// </summary>
+        public bool Point => _point;
+
+        /// <summary>
+        /// The coordinates in the cube.
+        /// </summary>
+        public IEnumerable<double> Coordinates => _coords;
+
+        /// <summary>
+        /// The lower left coordinates of the cube.
+        /// </summary>
+        public IEnumerable<double> LowerLeft => _coords.Take(Dimensions);
+        
+        /// <summary>
+        /// The upper right coordinates of the cube.
+        /// </summary>
+        public IEnumerable<double> UpperRight => _coords.Skip(Point ? 0 : Dimensions).Take(Dimensions);
+
+        /// <summary>
+        /// Checks if the cube is a point.
+        /// </summary>
+        /// <returns></returns>
+        internal bool IsPointInternal()
+        {
+            if (Point) return true;
+            
+            /*
+             * Even if the point-flag is not set, all the lower-left coordinates might
+             * match the upper-right coordinates, so that the value is in fact a
+             * point. Such values don't arise with current code - the point flag is
+             * always set if appropriate - but they might be present on-disk in
+             * clusters upgraded from pre-9.4 versions.
+             */
+            for (var i = 0; i < Dimensions; i++)
+            {
+                if (LowerLeft.ElementAt(i) != UpperRight.ElementAt(i)) return false;
+            }
+            return true;
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            var builder = new StringBuilder();
+
+            builder.Append('(');
+            for (var i = 0; i < Dimensions; i++)
+            {
+                if (i > 0) builder.Append(", ");
+                builder.Append(LowerLeft.ElementAt(i));
+            }
+            builder.Append(')');
+
+            if (!IsPointInternal())
+            {
+                builder.Append(",(");
+                for (var i = 0; i < Dimensions; i++)
+                {
+                    if (i > 0) builder.Append(", ");
+                    builder.Append(UpperRight.ElementAt(i));
+                }
+                builder.Append(')');
+            }
+
+            return builder.ToString();
+        }
+
+        public bool Equals(NpgsqlCube other) => 
+            Dimensions == other.Dimensions && Coordinates.SequenceEqual(other.Coordinates) && Point == other.Point;
+
+        public override bool Equals(object? obj) => obj is NpgsqlCube other && Equals(other);
+
+        public static bool operator ==(NpgsqlCube x, NpgsqlCube y) => x.Equals(y);
+
+        public static bool operator !=(NpgsqlCube x, NpgsqlCube y) => !(x == y);
+        
+        public override int GetHashCode() => 
+            Dimensions.GetHashCode() * Coordinates.GetHashCode() * Point.GetHashCode();
```

Comment:
Use HashCode.Combine.

Also, does Dimensions need to be included given that it's calculated from Point and Coordinates?

Reply:
Probably not, honestly looking at this with fresh eyes, we probably only need to get the hash code of the lower-left and upper-right coordinates and if they point to the same double values, then that will determine if it's a point or not.